# Pruebas A/B
## Introducción y objetivos

En este proyecto, se comenzó a realizar una tarea analítica de una tienda en línea internacional en la que se usará una **prueba A/B** en la que el grupo A es el *control* y el grupo B es el *embudo nuevo de pago.* El propósito de la prueba es probar cambios relacionados con la introducción de un sistema mejorado de recomendaciones, se espera que alrededor de 6000 participantes de la prueba, así como se espera que dentro de *14 días* posteriores a la inscripción, los usuarios logren mostrar una mejor conversión en vistas de la página del producto, instancias de agregar artículos al carrito de compras y compras. En cada etapa del embudo **product_page → product_cart → purchase,** habrá al menos un 10% de aumento.

## Exploración y análisis exploratorio de datos

### Información general

Para iniciar se importaron las liberias recomendadas, y se leyeron los *datasets* que se tienen para revisar si hay anomalías

In [1]:
import pandas as pd

In [2]:
ev_cal = pd.read_csv('/datasets/ab_project_marketing_events_us.csv') 
new_users = pd.read_csv('/datasets/final_ab_new_users_upd_us.csv')
upd_ev = pd.read_csv('final_ab_events_upd_us.csv')
part = pd.read_csv('final_ab_participants_upd_us.csv')

In [3]:
print("=== INFORMACIÓN GENERAL ===")

print("\nInformación general")
ev_cal.info()

print("\nPrimeras filas:")
print(ev_cal.head())

print("\n=== TIPOS DE DATOS ===")
print(ev_cal.dtypes)

print("\nDescripción")
ev_cal.describe()
print()
print(f"Forma del dataset: {ev_cal.shape}")

=== INFORMACIÓN GENERAL ===

Información general
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       14 non-null     object
 1   regions    14 non-null     object
 2   start_dt   14 non-null     object
 3   finish_dt  14 non-null     object
dtypes: object(4)
memory usage: 576.0+ bytes

Primeras filas:
                           name                   regions    start_dt  \
0      Christmas&New Year Promo             EU, N.America  2020-12-25   
1  St. Valentine's Day Giveaway  EU, CIS, APAC, N.America  2020-02-14   
2        St. Patric's Day Promo             EU, N.America  2020-03-17   
3                  Easter Promo  EU, CIS, APAC, N.America  2020-04-12   
4             4th of July Promo                 N.America  2020-07-04   

    finish_dt  
0  2021-01-03  
1  2020-02-16  
2  2020-03-19  
3  2020-04-19  
4  2020-07-11  

=== TIPOS DE DATO

In [4]:
print("=== INFORMACIÓN GENERAL ===")

print("\nInformación general")
new_users.info()

print("\nPrimeras filas:")
print(new_users.head())

print("\n=== TIPOS DE DATOS ===")
print(new_users.dtypes)

print("\nDescripción")
new_users.describe()
print()
print(f"Forma del dataset: {new_users.shape}")

=== INFORMACIÓN GENERAL ===

Información general
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58703 entries, 0 to 58702
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     58703 non-null  object
 1   first_date  58703 non-null  object
 2   region      58703 non-null  object
 3   device      58703 non-null  object
dtypes: object(4)
memory usage: 1.8+ MB

Primeras filas:
            user_id  first_date     region   device
0  D72A72121175D8BE  2020-12-07         EU       PC
1  F1C668619DFE6E65  2020-12-07  N.America  Android
2  2E1BF1D4C37EA01F  2020-12-07         EU       PC
3  50734A22C0C63768  2020-12-07         EU   iPhone
4  E1BDDCE0DAFA2679  2020-12-07  N.America   iPhone

=== TIPOS DE DATOS ===
user_id       object
first_date    object
region        object
device        object
dtype: object

Descripción

Forma del dataset: (58703, 4)


In [5]:
print("\n=== INFORMACIÓN GENERAL ===")

print("\nInformación general")
upd_ev.info()

print("\nPrimeras filas:")
print(upd_ev.head())

print("\n=== TIPOS DE DATOS ===")
print(upd_ev.dtypes)

print("\nDescripción")
upd_ev.describe()
print()
print(f"Forma del dataset: {upd_ev.shape}")


=== INFORMACIÓN GENERAL ===

Información general
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423761 entries, 0 to 423760
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   user_id     423761 non-null  object 
 1   event_dt    423761 non-null  object 
 2   event_name  423761 non-null  object 
 3   details     60314 non-null   float64
dtypes: float64(1), object(3)
memory usage: 12.9+ MB

Primeras filas:
            user_id             event_dt event_name  details
0  E1BDDCE0DAFA2679  2020-12-07 20:22:03   purchase    99.99
1  7B6452F081F49504  2020-12-07 09:22:53   purchase     9.99
2  9CD9F34546DF254C  2020-12-07 12:59:29   purchase     4.99
3  96F27A054B191457  2020-12-07 04:02:40   purchase     4.99
4  1FD7660FDF94CA1F  2020-12-07 10:15:09   purchase     4.99

=== TIPOS DE DATOS ===
user_id        object
event_dt       object
event_name     object
details       float64
dtype: object

Descripción

Forma del da

In [6]:
print("\n=== INFORMACIÓN GENERAL ===")

print("\nInformación general")
part.info()

print("\nPrimeras filas:")
print(part.head())

print("\n=== TIPOS DE DATOS ===")
print(part.dtypes)

print("\nDescripción")
part.describe()
print()
print(f"Forma del dataset: {part.shape}")


=== INFORMACIÓN GENERAL ===

Información general
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14525 entries, 0 to 14524
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  14525 non-null  object
 1   group    14525 non-null  object
 2   ab_test  14525 non-null  object
dtypes: object(3)
memory usage: 340.6+ KB

Primeras filas:
            user_id group                  ab_test
0  D1ABA3E2887B6A73     A  recommender_system_test
1  A7A3664BD6242119     A  recommender_system_test
2  DABC14FDDFADD29E     A  recommender_system_test
3  04988C5DF189632E     A  recommender_system_test
4  4FF2998A348C484F     A  recommender_system_test

=== TIPOS DE DATOS ===
user_id    object
group      object
ab_test    object
dtype: object

Descripción

Forma del dataset: (14525, 3)


### Conversión y limpieza

Tras observar en *"Líneas generales"* la información de cada tabla, se procedió a identificar los datos duplicados y/o ausentes, para posteriormente eliminarlos y/o reemplazarlos

#### Valores Ausentes

In [7]:
print("=== VALORES AUSENTES ===")
missing_data = ev_cal.isnull().sum()
missing_percent = (ev_cal.isnull().sum() / len(ev_cal)) * 100

missing_summary = pd.DataFrame({
    'Valores_Ausentes': missing_data,
    'Porcentaje': missing_percent
})

if missing_data.sum() == 0:
    print("¡Excelente! No hay valores ausentes")
else:
    print("Se encontraron valores ausentes que necesitan atención")

print(missing_summary)
print()

=== VALORES AUSENTES ===
¡Excelente! No hay valores ausentes
           Valores_Ausentes  Porcentaje
name                      0         0.0
regions                   0         0.0
start_dt                  0         0.0
finish_dt                 0         0.0



In [8]:
print("=== VALORES AUSENTES ===")
missing_data = new_users.isnull().sum()
missing_percent = (new_users.isnull().sum() / len(new_users)) * 100

missing_summary = pd.DataFrame({
    'Valores_Ausentes': missing_data,
    'Porcentaje': missing_percent
})

if missing_data.sum() == 0:
    print("¡Excelente! No hay valores ausentes")
else:
    print("Se encontraron valores ausentes que necesitan atención")

print(missing_summary)
print()

=== VALORES AUSENTES ===
¡Excelente! No hay valores ausentes
            Valores_Ausentes  Porcentaje
user_id                    0         0.0
first_date                 0         0.0
region                     0         0.0
device                     0         0.0



In [9]:
print("=== VALORES AUSENTES ===")
missing_data = upd_ev.isnull().sum()
missing_percent = (upd_ev.isnull().sum() / len(upd_ev)) * 100

missing_summary = pd.DataFrame({
    'Valores_Ausentes': missing_data,
    'Porcentaje': missing_percent
})

if missing_data.sum() == 0:
    print("¡Excelente! No hay valores ausentes")
else:
    print("Se encontraron valores ausentes que necesitan atención")

print(missing_summary)
print()

=== VALORES AUSENTES ===
Se encontraron valores ausentes que necesitan atención
            Valores_Ausentes  Porcentaje
user_id                    0    0.000000
event_dt                   0    0.000000
event_name                 0    0.000000
details               363447   85.766977



In [10]:
print("=== VALORES AUSENTES ===")
missing_data = part.isnull().sum()
missing_percent = (part.isnull().sum() / len(part)) * 100

missing_summary = pd.DataFrame({
    'Valores_Ausentes': missing_data,
    'Porcentaje': missing_percent
})

if missing_data.sum() == 0:
    print("¡Excelente! No hay valores ausentes")
else:
    print("Se encontraron valores ausentes que necesitan atención")

print(missing_summary)
print()

=== VALORES AUSENTES ===
¡Excelente! No hay valores ausentes
         Valores_Ausentes  Porcentaje
user_id                 0         0.0
group                   0         0.0
ab_test                 0         0.0



In [11]:
usuarios_duplicados = part[part.duplicated('user_id', keep=False)]
participants = part.drop_duplicates('user_id', keep='first')

#### Valores duplicados

In [12]:
print("=== DUPLICADOS ===")
duplicates = ev_cal.duplicated().sum()
print(f"Filas duplicadas: {duplicates:,}")

if duplicates > 0:
    print("Se encontraron filas duplicadas")
    # Mostrar algunos ejemplos
    print("Ejemplos de duplicados:")
    print(ev_cal[ev_cal.duplicated()].head())
else:
    print("No hay filas duplicadas")

=== DUPLICADOS ===
Filas duplicadas: 0
No hay filas duplicadas


In [13]:
print("=== DUPLICADOS ===")
duplicates = new_users.duplicated().sum()
print(f"Filas duplicadas: {duplicates:,}")

if duplicates > 0:
    print("Se encontraron filas duplicadas")
    # Mostrar algunos ejemplos
    print("Ejemplos de duplicados:")
    print(new_users[new_users.duplicated()].head())
else:
    print("No hay filas duplicadas")

=== DUPLICADOS ===
Filas duplicadas: 0
No hay filas duplicadas


In [14]:
print("=== DUPLICADOS ===")
duplicates = upd_ev.duplicated().sum()
print(f"Filas duplicadas: {duplicates:,}")

if duplicates > 0:
    print("Se encontraron filas duplicadas")
    # Mostrar algunos ejemplos
    print("Ejemplos de duplicados:")
    print(upd_ev[upd_ev.duplicated()].head())
else:
    print("No hay filas duplicadas")

=== DUPLICADOS ===
Filas duplicadas: 0
No hay filas duplicadas


In [15]:
print("=== DUPLICADOS ===")
duplicates = participants.duplicated().sum()
print(f"Filas duplicadas: {duplicates:,}")

if duplicates > 0:
    print("Se encontraron filas duplicadas")
    # Mostrar algunos ejemplos
    print("Ejemplos de duplicados:")
    print(participants[participants.duplicated()].head())
else:
    print("No hay filas duplicadas")

=== DUPLICADOS ===
Filas duplicadas: 0
No hay filas duplicadas


#### Datos de usuarios únicos

In [16]:
print("=== VALORES ÚNICOS ===")
for column in ev_cal.columns:
    unique_count = ev_cal[column].nunique()
    print(f"{column}: {unique_count:,} valores únicos")
    
    # Mostrar algunos ejemplos para columnas categóricas
    if column in ['EventName', 'ExpId']:
        print(f"  Valores: {sorted(ev_cal[column].unique())}")
    print()

=== VALORES ÚNICOS ===
name: 14 valores únicos

regions: 6 valores únicos

start_dt: 14 valores únicos

finish_dt: 14 valores únicos



In [17]:
print("=== VALORES ÚNICOS ===")
for column in new_users.columns:
    unique_count = new_users[column].nunique()
    print(f"{column}: {unique_count:,} valores únicos")
    
    # Mostrar algunos ejemplos para columnas categóricas
    if column in ['EventName', 'ExpId']:
        print(f"  Valores: {sorted(new_users[column].unique())}")
    print()

=== VALORES ÚNICOS ===
user_id: 58,703 valores únicos

first_date: 17 valores únicos

region: 4 valores únicos

device: 4 valores únicos



In [18]:
print("=== VALORES ÚNICOS ===")
for column in upd_ev.columns:
    unique_count = upd_ev[column].nunique()
    total_rows = len(upd_ev)
    print(f"\n{column.upper()}:")
    print(f"  Valores únicos: {unique_count:,}")
    print(f"  Total filas: {total_rows:,}")
    print(f"  Ratio únicos/total: {unique_count/total_rows:.3f}")
    
    # Mostrar valores únicos para columnas específicas
    if column == 'event_name':
        print(f"  Tipos de eventos: {sorted(upd_ev[column].unique())}")
        print(f"  Distribución:")
        print(upd_ev[column].value_counts().to_string())
    elif column == 'details':
        print(f"  Valores únicos de details: {sorted(upd_ev[column].dropna().unique())}")

=== VALORES ÚNICOS ===

USER_ID:
  Valores únicos: 58,703
  Total filas: 423,761
  Ratio únicos/total: 0.139

EVENT_DT:
  Valores únicos: 257,138
  Total filas: 423,761
  Ratio únicos/total: 0.607

EVENT_NAME:
  Valores únicos: 4
  Total filas: 423,761
  Ratio únicos/total: 0.000
  Tipos de eventos: ['login', 'product_cart', 'product_page', 'purchase']
  Distribución:
login           182465
product_page    120862
purchase         60314
product_cart     60120

DETAILS:
  Valores únicos: 4
  Total filas: 423,761
  Ratio únicos/total: 0.000
  Valores únicos de details: [4.99, 9.99, 99.99, 499.99]


In [19]:
print("=== VALORES ÚNICOS ===")
for column in participants.columns:
    unique_count = participants[column].nunique()
    total_rows = len(part)
    print(f"\n{column.upper()}:")
    print(f"  Valores únicos: {unique_count:,}")
    print(f"  Total filas: {total_rows:,}")
    print(f"  Ratio únicos/total: {unique_count/total_rows:.3f}")
    
    # Mostrar valores únicos para columnas categóricas
    if column in ['group', 'ab_test']:
        print(f"  Valores: {sorted(participants[column].unique())}")
        print(f"  Distribución:")
        print(participants[column].value_counts().to_string())

=== VALORES ÚNICOS ===

USER_ID:
  Valores únicos: 13,638
  Total filas: 14,525
  Ratio únicos/total: 0.939

GROUP:
  Valores únicos: 2
  Total filas: 14,525
  Ratio únicos/total: 0.000
  Valores: ['A', 'B']
  Distribución:
A    7758
B    5880

AB_TEST:
  Valores únicos: 2
  Total filas: 14,525
  Ratio únicos/total: 0.000
  Valores: ['interface_eu_test', 'recommender_system_test']
  Distribución:
interface_eu_test          9963
recommender_system_test    3675


In [20]:
print("=== VERIFICACIÓN DETALLADA DE USUARIOS ÚNICOS - PARTICIPANTES ===")

# Información básica
total_filas = len(part)
usuarios_unicos = part['user_id'].nunique()
usuarios_duplicados_count = total_filas - usuarios_unicos

print(f"Total de filas en part: {total_filas:,}")
print(f"Usuarios únicos: {usuarios_unicos:,}")
print(f"Usuarios duplicados: {usuarios_duplicados_count:,}")
print(f"Ratio únicos/total: {usuarios_unicos/total_filas:.3f}")

# Verificar duplicados específicos
if usuarios_duplicados_count > 0:
    duplicados = part[part.duplicated('user_id', keep=False)]
    print(f"\n🚨 PROBLEMA: {usuarios_duplicados_count} usuarios aparecen múltiples veces")
    print("Primeros 5 usuarios duplicados:")
    print(duplicados.head())

=== VERIFICACIÓN DETALLADA DE USUARIOS ÚNICOS - PARTICIPANTES ===
Total de filas en part: 14,525
Usuarios únicos: 13,638
Usuarios duplicados: 887
Ratio únicos/total: 0.939

🚨 PROBLEMA: 887 usuarios aparecen múltiples veces
Primeros 5 usuarios duplicados:
             user_id group                  ab_test
2   DABC14FDDFADD29E     A  recommender_system_test
3   04988C5DF189632E     A  recommender_system_test
8   B3A2485649E4A012     A  recommender_system_test
15  EAFB9027A27D510C     B  recommender_system_test
17  5D5E6EE92AF6E9E0     B  recommender_system_test


En el *dataset* de los eventos de nuevos usuarios (upd_ev) se han encontrado **datos ausentes**, asi como todos los *dataset* tienen el formato de **fecha incorrecto**, incluso, el *dataset* de (part) posee datos de usuarios duplicados, asi que se procedió a cambiar los formatos adecuados

#### Conversión de tipos

Reemplazo de *object* a *datetime* en columnas de fecha y creación de variables categóricas

In [21]:
# Conversación de datos

ev_cal['start_dt'] = pd.to_datetime(ev_cal['start_dt'])

ev_cal['finish_dt'] = pd.to_datetime(ev_cal['finish_dt'])

new_users['first_date'] = pd.to_datetime(new_users['first_date'])

upd_ev['event_dt'] = pd.to_datetime(upd_ev['event_dt'])

# Conversión de categorías
new_users['region'] = new_users['region'].astype('category')
new_users['device'] = new_users['device'].astype('category')

upd_ev['event_name'] = upd_ev['event_name'].astype('category')

#### Verificación crítica del dataset

Debido a que uno de los dataset se encontró con el problema de datos duplicados, se decidió a realizar un análisis para su posible resolución

In [22]:
print("\n=== VERIFICACIÓN CRUZADA ENTRE DATASETS ===")

# Usuarios únicos en cada dataset
usuarios_part = set(part['user_id'].unique())
usuarios_upd_ev = set(upd_ev['user_id'].unique())
usuarios_new_users = set(new_users['user_id'].unique())

print(f"Usuarios únicos en part: {len(usuarios_part):,}")
print(f"Usuarios únicos en upd_ev: {len(usuarios_upd_ev):,}")
print(f"Usuarios únicos en new_users: {len(usuarios_new_users):,}")

# Intersecciones
part_en_eventos = usuarios_part.intersection(usuarios_upd_ev)
part_en_new_users = usuarios_part.intersection(usuarios_new_users)

print(f"\nUsuarios de part que están en upd_ev: {len(part_en_eventos):,}")
print(f"Usuarios de part que están en new_users: {len(part_en_new_users):,}")

# Usuarios que faltan
part_no_en_eventos = usuarios_part - usuarios_upd_ev
part_no_en_new_users = usuarios_part - usuarios_new_users

if len(part_no_en_eventos) > 0:
    print(f"🚨 Usuarios de part SIN eventos: {len(part_no_en_eventos):,}")
if len(part_no_en_new_users) > 0:
    print(f"🚨 Usuarios de part NO en new_users: {len(part_no_en_new_users):,}")


=== VERIFICACIÓN CRUZADA ENTRE DATASETS ===
Usuarios únicos en part: 13,638
Usuarios únicos en upd_ev: 58,703
Usuarios únicos en new_users: 58,703

Usuarios de part que están en upd_ev: 13,638
Usuarios de part que están en new_users: 13,638


In [23]:
print("\n=== VERIFICACIÓN DE USUARIOS EN AMBOS GRUPOS ===")

# Usuarios por grupo
usuarios_grupo_A = set(part[part['group'] == 'A']['user_id'])
usuarios_grupo_B = set(part[part['group'] == 'B']['user_id'])

# Intersección (usuarios en ambos grupos)
usuarios_en_ambos = usuarios_grupo_A.intersection(usuarios_grupo_B)

print(f"Usuarios únicos en Grupo A: {len(usuarios_grupo_A):,}")
print(f"Usuarios únicos en Grupo B: {len(usuarios_grupo_B):,}")
print(f"Usuarios en AMBOS grupos: {len(usuarios_en_ambos):,}")

if len(usuarios_en_ambos) > 0:
    print(f"\n🚨 PROBLEMA CRÍTICO: {len(usuarios_en_ambos)} usuarios están en ambos grupos")
    print("Esto viola completamente la independencia entre grupos")
    
    # Mostrar algunos ejemplos
    print(f"\nPrimeros 10 usuarios duplicados:")
    for i, user_id in enumerate(list(usuarios_en_ambos)[:10]):
        print(f"  {i+1}. {user_id}")
        
    # Verificar asignaciones de estos usuarios
    print(f"\nEjemplo de asignaciones duplicadas:")
    ejemplo_user = list(usuarios_en_ambos)[0]
    asignaciones = part[part['user_id'] == ejemplo_user]
    print(asignaciones)
else:
    print("✅ Correcto: No hay usuarios duplicados entre grupos")


=== VERIFICACIÓN DE USUARIOS EN AMBOS GRUPOS ===
Usuarios únicos en Grupo A: 7,874
Usuarios únicos en Grupo B: 6,205
Usuarios en AMBOS grupos: 441

🚨 PROBLEMA CRÍTICO: 441 usuarios están en ambos grupos
Esto viola completamente la independencia entre grupos

Primeros 10 usuarios duplicados:
  1. 0BA00E790AA510C1
  2. B06B45B7C3E5E721
  3. CFDED9167B27A57F
  4. BBE0583F0240AABB
  5. 2CA426D092054330
  6. 4CCE583A34A8362C
  7. FAA7192A42F9F8DA
  8. 12A77BCF101643B5
  9. 9270DB42112E3B14
  10. F71DD44DAB783CEB

Ejemplo de asignaciones duplicadas:
                user_id group                  ab_test
3226   0BA00E790AA510C1     A  recommender_system_test
10564  0BA00E790AA510C1     B        interface_eu_test


In [24]:
print("\n=== ANÁLISIS DEL IMPACTO ===")

# Calcular el porcentaje de contaminación
total_usuarios_unicos = len(usuarios_grupo_A.union(usuarios_grupo_B))
porcentaje_contaminacion = (len(usuarios_en_ambos) / total_usuarios_unicos) * 100

print(f"Total usuarios únicos reales: {total_usuarios_unicos:,}")
print(f"Usuarios contaminados: {len(usuarios_en_ambos):,}")
print(f"Porcentaje de contaminación: {porcentaje_contaminacion:.1f}%")

# Impacto en el tamaño efectivo de los grupos
usuarios_solo_A = usuarios_grupo_A - usuarios_en_ambos
usuarios_solo_B = usuarios_grupo_B - usuarios_en_ambos

print(f"\nTamaños efectivos sin contaminación:")
print(f"Solo en Grupo A: {len(usuarios_solo_A):,}")
print(f"Solo en Grupo B: {len(usuarios_solo_B):,}")
print(f"En ambos grupos: {len(usuarios_en_ambos):,}")


=== ANÁLISIS DEL IMPACTO ===
Total usuarios únicos reales: 13,638
Usuarios contaminados: 441
Porcentaje de contaminación: 3.2%

Tamaños efectivos sin contaminación:
Solo en Grupo A: 7,433
Solo en Grupo B: 5,764
En ambos grupos: 441


In [25]:
part_limpio = part.drop_duplicates('user_id', keep='first')

print(f"Datos originales: {len(part):,} filas")
print(f"Datos limpios: {len(part_limpio):,} filas")
print(f"Filas eliminadas: {len(part) - len(part_limpio):,}")

Datos originales: 14,525 filas
Datos limpios: 13,638 filas
Filas eliminadas: 887


In [26]:
# Limpieza de datos
print("=== Limpieza de datos ===")

# Crear dataset limpio eliminando duplicados
part_limpio = part.drop_duplicates('user_id', keep='first')

# Verificar la limpieza
print(f"Datos originales: {len(part):,} filas")
print(f"Datos limpios: {len(part_limpio):,} filas") 
print(f"Filas eliminadas: {len(part) - len(part_limpio):,}")

# Verificar que no hay más duplicados
usuarios_A_limpio = set(part_limpio[part_limpio['group'] == 'A']['user_id'])
usuarios_B_limpio = set(part_limpio[part_limpio['group'] == 'B']['user_id'])
duplicados_restantes = usuarios_A_limpio.intersection(usuarios_B_limpio)

print(f"\n=== VERIFICACIÓN POST-LIMPIEZA ===")
print(f"Usuarios en Grupo A: {len(usuarios_A_limpio):,}")
print(f"Usuarios en Grupo B: {len(usuarios_B_limpio):,}")
print(f"Usuarios duplicados restantes: {len(duplicados_restantes):,}")

=== Limpieza de datos ===
Datos originales: 14,525 filas
Datos limpios: 13,638 filas
Filas eliminadas: 887

=== VERIFICACIÓN POST-LIMPIEZA ===
Usuarios en Grupo A: 7,758
Usuarios en Grupo B: 5,880
Usuarios duplicados restantes: 0


### Conversión de etapas del embudo

Para realizarlo, se realizó la tasa de conversión para identificar que usuarios pasan de una etapa a otra, primeramente, se idetificó a los uusarios únicos en cada etapa del embudo por grupo

In [27]:
print("=== USUARIOS ÚNICOS EN CADA ETAPA DEL EMBUDO POR GRUPO ===")

# Unir datos de eventos con participantes limpios
eventos_ab_limpio = upd_ev.merge(part_limpio, on='user_id', how='inner')

# Calcular usuarios únicos por grupo y etapa del embudo
embudo_por_grupo_limpio = eventos_ab_limpio.groupby(['group', 'event_name'])['user_id'].nunique().unstack(fill_value=0)

print("USUARIOS ÚNICOS POR ETAPA Y GRUPO:")
print(embudo_por_grupo_limpio)

# Extraer solo las etapas del embudo objetivo: product_page → product_cart → purchase
etapas_embudo = ['product_page', 'product_cart', 'purchase']
embudo_objetivo = embudo_por_grupo_limpio[etapas_embudo]

print(f"\n=== EMBUDO OBJETIVO: product_page → product_cart → purchase ===")
print(embudo_objetivo)

=== USUARIOS ÚNICOS EN CADA ETAPA DEL EMBUDO POR GRUPO ===
USUARIOS ÚNICOS POR ETAPA Y GRUPO:
event_name  login  product_cart  product_page  purchase
group                                                  
A            7757          2453          5146      2654
B            5879          1941          3777      1909

=== EMBUDO OBJETIVO: product_page → product_cart → purchase ===
event_name  product_page  product_cart  purchase
group                                           
A                   5146          2453      2654
B                   3777          1941      1909


In [28]:
print("\n=== ANÁLISIS DETALLADO POR GRUPO ===")

# Grupo A (Control)
grupo_A = embudo_objetivo.loc['A']
print("GRUPO A (Control - Sistema Actual):")
print(f"Product Page: {grupo_A['product_page']:,} usuarios")
print(f"Product Cart: {grupo_A['product_cart']:,} usuarios")
print(f"Purchase: {grupo_A['purchase']:,} usuarios")

# Grupo B (Tratamiento)
grupo_B = embudo_objetivo.loc['B']
print("\nGRUPO B (Tratamiento - Sistema Mejorado):")
print(f"Product Page: {grupo_B['product_page']:,} usuarios")
print(f"Product Cart: {grupo_B['product_cart']:,} usuarios")
print(f"Purchase: {grupo_B['purchase']:,} usuarios")

# Comparación directa
print(f"\n=== COMPARACIÓN DIRECTA ===")
print(f"Diferencia en Product Page: {grupo_B['product_page'] - grupo_A['product_page']:+,} usuarios")
print(f"Diferencia en Product Cart: {grupo_B['product_cart'] - grupo_A['product_cart']:+,} usuarios")
print(f"Diferencia en Purchase: {grupo_B['purchase'] - grupo_A['purchase']:+,} usuarios")


=== ANÁLISIS DETALLADO POR GRUPO ===
GRUPO A (Control - Sistema Actual):
Product Page: 5,146 usuarios
Product Cart: 2,453 usuarios
Purchase: 2,654 usuarios

GRUPO B (Tratamiento - Sistema Mejorado):
Product Page: 3,777 usuarios
Product Cart: 1,941 usuarios
Purchase: 1,909 usuarios

=== COMPARACIÓN DIRECTA ===
Diferencia en Product Page: -1,369 usuarios
Diferencia en Product Cart: -512 usuarios
Diferencia en Purchase: -745 usuarios


In [29]:
print(f"\n=== PORCENTAJES DE PARTICIPACIÓN ===")

# Total de usuarios por grupo (datos limpios)
total_A = len(usuarios_A_limpio)
total_B = len(usuarios_B_limpio)

print("GRUPO A (Control):")
print(f"  Total usuarios: {total_A:,}")
print(f"  Vieron páginas: {grupo_A['product_page']:,} ({grupo_A['product_page']/total_A*100:.1f}%)")
print(f"  Agregaron al carrito: {grupo_A['product_cart']:,} ({grupo_A['product_cart']/total_A*100:.1f}%)")
print(f"  Realizaron compras: {grupo_A['purchase']:,} ({grupo_A['purchase']/total_A*100:.1f}%)")

print("\nGRUPO B (Tratamiento):")
print(f"  Total usuarios: {total_B:,}")
print(f"  Vieron páginas: {grupo_B['product_page']:,} ({grupo_B['product_page']/total_B*100:.1f}%)")
print(f"  Agregaron al carrito: {grupo_B['product_cart']:,} ({grupo_B['product_cart']/total_B*100:.1f}%)")
print(f"  Realizaron compras: {grupo_B['purchase']:,} ({grupo_B['purchase']/total_B*100:.1f}%)")


=== PORCENTAJES DE PARTICIPACIÓN ===
GRUPO A (Control):
  Total usuarios: 7,758
  Vieron páginas: 5,146 (66.3%)
  Agregaron al carrito: 2,453 (31.6%)
  Realizaron compras: 2,654 (34.2%)

GRUPO B (Tratamiento):
  Total usuarios: 5,880
  Vieron páginas: 3,777 (64.2%)
  Agregaron al carrito: 1,941 (33.0%)
  Realizaron compras: 1,909 (32.5%)


In [30]:

print(f"\n=== TASAS DE CONVERSIÓN ENTRE ETAPAS ===")

# Grupo A
conv_A_page_cart = (grupo_A['product_cart'] / grupo_A['product_page']) * 100
conv_A_cart_purchase = (grupo_A['purchase'] / grupo_A['product_cart']) * 100
conv_A_page_purchase = (grupo_A['purchase'] / grupo_A['product_page']) * 100

print("GRUPO A (Control):")
print(f"  Page → Cart: {conv_A_page_cart:.1f}%")
print(f"  Cart → Purchase: {conv_A_cart_purchase:.1f}%")
print(f"  Page → Purchase: {conv_A_page_purchase:.1f}%")

# Grupo B
conv_B_page_cart = (grupo_B['product_cart'] / grupo_B['product_page']) * 100
conv_B_cart_purchase = (grupo_B['purchase'] / grupo_B['product_cart']) * 100
conv_B_page_purchase = (grupo_B['purchase'] / grupo_B['product_page']) * 100

print("\nGRUPO B (Tratamiento):")
print(f"  Page → Cart: {conv_B_page_cart:.1f}%")
print(f"  Cart → Purchase: {conv_B_cart_purchase:.1f}%")
print(f"  Page → Purchase: {conv_B_page_purchase:.1f}%")

# Diferencias
print(f"\n=== DIFERENCIAS EN CONVERSIÓN ===")
print(f"Page → Cart: {conv_B_page_cart - conv_A_page_cart:+.1f} puntos porcentuales")
print(f"Cart → Purchase: {conv_B_cart_purchase - conv_A_cart_purchase:+.1f} puntos porcentuales")
print(f"Page → Purchase: {conv_B_page_purchase - conv_A_page_purchase:+.1f} puntos porcentuales")



=== TASAS DE CONVERSIÓN ENTRE ETAPAS ===
GRUPO A (Control):
  Page → Cart: 47.7%
  Cart → Purchase: 108.2%
  Page → Purchase: 51.6%

GRUPO B (Tratamiento):
  Page → Cart: 51.4%
  Cart → Purchase: 98.4%
  Page → Purchase: 50.5%

=== DIFERENCIAS EN CONVERSIÓN ===
Page → Cart: +3.7 puntos porcentuales
Cart → Purchase: -9.8 puntos porcentuales
Page → Purchase: -1.0 puntos porcentuales


In [31]:
print("=== ANÁLISIS DE PÉRDIDA DE USUARIOS ===")

# Grupo A
perdida_A_page_cart = grupo_A['product_page'] - grupo_A['product_cart']
perdida_A_cart_purchase = grupo_A['product_cart'] - grupo_A['purchase']

print("GRUPO A (Control):")
print(f"  Pérdida Page → Cart: {perdida_A_page_cart:,} usuarios ({perdida_A_page_cart/grupo_A['product_page']*100:.1f}%)")
print(f"  Pérdida Cart → Purchase: {perdida_A_cart_purchase:,} usuarios ({abs(perdida_A_cart_purchase)/grupo_A['product_cart']*100:.1f}%)")

# Grupo B
perdida_B_page_cart = grupo_B['product_page'] - grupo_B['product_cart']
perdida_B_cart_purchase = grupo_B['product_cart'] - grupo_B['purchase']

print("\nGRUPO B (Tratamiento):")
print(f"  Pérdida Page → Cart: {perdida_B_page_cart:,} usuarios ({perdida_B_page_cart/grupo_B['product_page']*100:.1f}%)")
print(f"  Pérdida Cart → Purchase: {perdida_B_cart_purchase:,} usuarios ({abs(perdida_B_cart_purchase)/grupo_B['product_cart']*100:.1f}%)")

=== ANÁLISIS DE PÉRDIDA DE USUARIOS ===
GRUPO A (Control):
  Pérdida Page → Cart: 2,693 usuarios (52.3%)
  Pérdida Cart → Purchase: -201 usuarios (8.2%)

GRUPO B (Tratamiento):
  Pérdida Page → Cart: 1,836 usuarios (48.6%)
  Pérdida Cart → Purchase: 32 usuarios (1.6%)


La mayor pérdida de usuarios ocurre en la etapa Page → Cart:

* Grupo A: Pierde 2,693 usuarios (52.3%)
* Grupo B: Pierde 1,836 usuarios (48.6%)

**La Anomalía Principal**
Se descubrió que la tasa de conversión de Cart → Purchase es del 108.2% en el Grupo A y 98.4% en el Grupo B. Esto significa que más usuarios compraron que los que agregaron productos al carrito, lo cual parece contradictorio.

¿Por qué ocurre esto?
La explicación principal es que los usuarios pueden comprar sin necesariamente agregar productos al carrito primero. Esto puede suceder por varias razones:

1. Compra Directa
* Algunos usuarios pueden tener la opción de "Comprar Ahora" que evita el carrito
* Productos con compra de un solo clic
* Usuarios que usan funciones de compra rápida

2. Comportamiento de Usuario Diverso
* Usuarios que navegan, salen, y regresan más tarde para comprar directamente
* Usuarios que usan diferentes dispositivos o sesiones
* Compras a través de enlaces directos o promociones

3. Limitaciones en el Tracking
* Posibles gaps en el seguimiento de eventos
* Usuarios que limpian cookies entre sesiones
* Diferentes flujos de compra no capturados completamente

### Enfoque de métricas más robustas

Debido a estas anomalías generadas en el estudio, se ha hecho cambios para seguir con las pruebas, entre los cambios, se implementaron  como: Enfoque en la Métrica Más Robusta, creación de muestra balanceada, recalculación de métricas con datos balanceados, y la aplicación de Prueba Z a Conversión Total

In [32]:
# Métricas más robustas

print(f"\n=== TASAS DE CONVERSIÓN ENTRE ETAPAS ===")

# Grupo A
conversion_A = grupo_A['purchase'] / grupo_A['product_page'] * 100 # 51.6%

# Grupo B
conversion_B = grupo_B['purchase'] / grupo_B['product_page'] * 100 # 50.5%

print("\nGRUPO A (Control):\n")
print(f"  Page → Purchase: {conversion_A:.1f}%")
print("\nGRUPO B (Tratamiento):\n")
print(f"  Page → Purchase: {conversion_B:.1f}%")


=== TASAS DE CONVERSIÓN ENTRE ETAPAS ===

GRUPO A (Control):

  Page → Purchase: 51.6%

GRUPO B (Tratamiento):

  Page → Purchase: 50.5%


In [33]:
import numpy as np
from scipy import stats

# Creación de muestra balanceada
print("=== CREANDO MUESTRA BALANCEADA ===")
np.random.seed(42)  # Para reproducibilidad

# Tomar el tamaño del grupo más pequeño
min_size = min(len(usuarios_A_limpio), len(usuarios_B_limpio))
print(f"Tamaño de muestra balanceada: {min_size:,} usuarios por grupo")

# Crear muestra aleatoria del Grupo A
usuarios_A_sample = np.random.choice(
    list(usuarios_A_limpio), 
    size=min_size, 
    replace=False
)

print(f"Grupo A original: {len(usuarios_A_limpio):,} usuarios")
print(f"Grupo A balanceado: {len(usuarios_A_sample):,} usuarios")
print(f"Grupo B: {len(usuarios_B_limpio):,} usuarios")

=== CREANDO MUESTRA BALANCEADA ===
Tamaño de muestra balanceada: 5,880 usuarios por grupo
Grupo A original: 7,758 usuarios
Grupo A balanceado: 5,880 usuarios
Grupo B: 5,880 usuarios


In [34]:
# Filtración de eventos para usuarios balanceados
print("\n=== RECALCULANDO CON DATOS BALANCEADOS ===")

# Lista de usuarios balanceados
usuarios_balanceados = list(usuarios_A_sample) + list(usuarios_B_limpio)

# Filtrar eventos solo para usuarios balanceados
eventos_balanceados = eventos_ab_limpio[
    eventos_ab_limpio['user_id'].isin(usuarios_balanceados)
]

# Recalcular embudo con datos balanceados
embudo_balanceado = eventos_balanceados.groupby(['group', 'event_name'])['user_id'].nunique().unstack(fill_value=0)
print("EMBUDO BALANCEADO:")
print(embudo_balanceado[['product_page', 'product_cart', 'purchase']])


=== RECALCULANDO CON DATOS BALANCEADOS ===
EMBUDO BALANCEADO:
event_name  product_page  product_cart  purchase
group                                           
A                   3912          1869      1990
B                   3777          1941      1909


## Prueba Z

Ahora que se han implementado la estrategia de rescate con datos balanceados, se hizo la prueba z a conversión total

In [40]:
# Extraer datos del embudo balanceado
grupo_A_bal = embudo_balanceado.loc['A']
grupo_B_bal = embudo_balanceado.loc['B']

# Datos para la prueba Z (Page → Purchase)
purchase_A_bal = grupo_A_bal['purchase']
page_A_bal = grupo_A_bal['product_page']
purchase_B_bal = grupo_B_bal['purchase']
page_B_bal = grupo_B_bal['product_page']

print(f"=== DATOS PARA PRUEBA Z BALANCEADA ===")
print(f"Grupo A: {purchase_A_bal}/{page_A_bal} = {purchase_A_bal/page_A_bal:.3f}")
print(f"Grupo B: {purchase_B_bal}/{page_B_bal} = {purchase_B_bal/page_B_bal:.3f}")

=== DATOS PARA PRUEBA Z BALANCEADA ===
Grupo A: 1990/3912 = 0.509
Grupo B: 1909/3777 = 0.505


In [39]:
import numpy as np
from scipy import stats

def prueba_z_proporciones(x1, n1, x2, n2):
    """
    x1, x2: número de éxitos en cada grupo
    n1, n2: tamaño de muestra de cada grupo
    """
    # Proporciones
    p1 = x1 / n1
    p2 = x2 / n2
    
    # Proporción combinada
    p_combined = (x1 + x2) / (n1 + n2)
    
    # Error estándar
    se = np.sqrt(p_combined * (1 - p_combined) * (1/n1 + 1/n2))
    
    # Estadístico z
    z = (p1 - p2) / se
    
    # Valor p (prueba de dos colas)
    p_value = 2 * (1 - stats.norm.cdf(abs(z)))
    
    return z, p_value, p1, p2

In [41]:
# Aplicar la prueba Z balanceada
z_stat_bal, p_value_bal, prop_A_bal, prop_B_bal = prueba_z_proporciones(
    purchase_A_bal, page_A_bal,
    purchase_B_bal, page_B_bal
)

print("\n=== RESULTADOS DE LA PRUEBA Z BALANCEADA ===")
print(f"Estadístico z: {z_stat_bal:.4f}")
print(f"Valor p: {p_value_bal:.6f}")
print(f"Proporción Grupo A: {prop_A_bal:.4f} ({prop_A_bal*100:.1f}%)")
print(f"Proporción Grupo B: {prop_B_bal:.4f} ({prop_B_bal*100:.1f}%)")
print(f"Diferencia: {(prop_A_bal - prop_B_bal)*100:.1f} puntos porcentuales")


=== RESULTADOS DE LA PRUEBA Z BALANCEADA ===
Estadístico z: 0.2862
Valor p: 0.774755
Proporción Grupo A: 0.5087 (50.9%)
Proporción Grupo B: 0.5054 (50.5%)
Diferencia: 0.3 puntos porcentuales


In [42]:
# Interpretación del resultado
alpha = 0.05
print(f"\n=== INTERPRETACIÓN (α = {alpha}) ===")
if p_value_bal < alpha:
    print("RESULTADO: Diferencia ESTADÍSTICAMENTE SIGNIFICATIVA")
    print("Rechazamos la hipótesis nula")
    print("El nuevo sistema de recomendaciones SÍ tiene un efecto significativo")
else:
    print("RESULTADO: Diferencia NO estadísticamente significativa")
    print("No podemos rechazar la hipótesis nula")
    print("No hay evidencia suficiente de que el nuevo sistema sea mejor")

print(f"\nConfianza: {(1-alpha)*100}%")


=== INTERPRETACIÓN (α = 0.05) ===
RESULTADO: Diferencia NO estadísticamente significativa
No podemos rechazar la hipótesis nula
No hay evidencia suficiente de que el nuevo sistema sea mejor

Confianza: 95.0%


In [43]:
# Calcular el tamaño del efecto
diferencia_absoluta = abs(prop_A_bal - prop_B_bal)
diferencia_relativa = (diferencia_absoluta / prop_A_bal) * 100

print(f"\n=== ANÁLISIS DE SIGNIFICANCIA PRÁCTICA ===")
print(f"Diferencia absoluta: {diferencia_absoluta:.4f} ({diferencia_absoluta*100:.1f} puntos porcentuales)")
print(f"Diferencia relativa: {diferencia_relativa:.1f}%")

# Tamaño del efecto (Cohen's h para proporciones)
import numpy as np

def cohens_h(p1, p2):
    """Calcula Cohen's h para diferencia entre proporciones"""
    return 2 * (np.arcsin(np.sqrt(p1)) - np.arcsin(np.sqrt(p2)))

h = abs(cohens_h(prop_A_bal, prop_B_bal))
print(f"Cohen's h (tamaño del efecto): {h:.4f}")

# Interpretación del tamaño del efecto
print(f"\n=== INTERPRETACIÓN DEL TAMAÑO DEL EFECTO ===")
if h < 0.2:
    efecto = "TRIVIAL"
elif h < 0.5:
    efecto = "PEQUEÑO"
elif h < 0.8:
    efecto = "MEDIANO"
else:
    efecto = "GRANDE"

print(f"Tamaño del efecto: {efecto}")


=== ANÁLISIS DE SIGNIFICANCIA PRÁCTICA ===
Diferencia absoluta: 0.0033 (0.3 puntos porcentuales)
Diferencia relativa: 0.6%
Cohen's h (tamaño del efecto): 0.0065

=== INTERPRETACIÓN DEL TAMAÑO DEL EFECTO ===
Tamaño del efecto: TRIVIAL


In [44]:
# Calcular el poder estadístico necesario para detectar esta diferencia
from scipy import stats

def calcular_tamano_muestra_necesario(p1, p2, alpha=0.05, power=0.8):
    """Calcula el tamaño de muestra necesario para detectar una diferencia"""
    effect_size = cohens_h(p1, p2)
    z_alpha = stats.norm.ppf(1 - alpha/2)
    z_beta = stats.norm.ppf(power)
    
    n = ((z_alpha + z_beta) / effect_size) ** 2
    return int(np.ceil(n))

n_necesario = calcular_tamano_muestra_necesario(prop_A_bal, prop_B_bal)
n_actual = min(page_A_bal, page_B_bal)

print(f"\n=== ANÁLISIS DE PODER ESTADÍSTICO ===")
print(f"Tamaño de muestra actual: {n_actual:,} por grupo")
print(f"Tamaño necesario para detectar esta diferencia: {n_necesario:,} por grupo")
print(f"Ratio actual/necesario: {n_actual/n_necesario:.2f}")

if n_actual < n_necesario:
    print("La muestra actual es INSUFICIENTE para detectar esta diferencia")
else:
    print("La muestra actual es suficiente")


=== ANÁLISIS DE PODER ESTADÍSTICO ===
Tamaño de muestra actual: 3,777 por grupo
Tamaño necesario para detectar esta diferencia: 184,188 por grupo
Ratio actual/necesario: 0.02
La muestra actual es INSUFICIENTE para detectar esta diferencia


### Distribución de eventos por día

Ahora se flitró una distribución de eventos por día y se comparó la activivad entre los grupos A y B


In [47]:
### Distribución de eventos por día
upd_ev['fecha'] = upd_ev['event_dt'].dt.date

# Actividad total por día
actividad_diaria = upd_ev.groupby('fecha').size()
print("=== ACTIVIDAD TOTAL POR DÍA ===")
print(actividad_diaria.head(10))


=== ACTIVIDAD TOTAL POR DÍA ===
fecha
2020-12-07    11385
2020-12-08    12547
2020-12-09    12122
2020-12-10    14077
2020-12-11    13864
2020-12-12    17634
2020-12-13    20985
2020-12-14    26184
2020-12-15    23469
2020-12-16    20909
dtype: int64


In [49]:
eventos_ab_limpio['fecha'] = eventos_ab_limpio['event_dt'].dt.date

# Actividad total por día y grupo
actividad_diaria_grupos = eventos_ab_limpio.groupby(['fecha', 'group']).size().unstack(fill_value=0)
print("=== ACTIVIDAD TOTAL POR DÍA Y GRUPO ===")
print(actividad_diaria_grupos.head(10))


=== ACTIVIDAD TOTAL POR DÍA Y GRUPO ===
group          A     B
fecha                 
2020-12-07  1286  1306
2020-12-08  1411  1350
2020-12-09  1469  1433
2020-12-10  1554  1458
2020-12-11  1605  1402
2020-12-12  2108  1810
2020-12-13  2276  2066
2020-12-14  3423  2400
2020-12-15  3136  2205
2020-12-16  3010  2107


### Interpretacíón de la significancia

Tras calcular la prueba Z, se procederá con la interpretación final de significancia práctica


In [45]:
print("=== INTERPRETACIÓN FINAL DE SIGNIFICANCIA PRÁCTICA ===")

# Resumen de resultados clave
print("📊 RESULTADOS CLAVE:")
print(f"• Diferencia observada: {(prop_A_bal - prop_B_bal)*100:.1f} puntos porcentuales")
print(f"• Tamaño del efecto (Cohen's h): {h:.4f} (TRIVIAL)")
print(f"• Significancia estadística: NO (p = {p_value_bal:.3f})")
print(f"• Muestra necesaria: {n_necesario:,} usuarios por grupo")
print(f"• Muestra actual: {n_actual:,} usuarios por grupo")

print("\n🎯 SIGNIFICANCIA PRÁCTICA:")
# Criterios de significancia práctica
umbral_practico = 0.05  # 5% de mejora mínima esperada
mejora_observada = abs(prop_A_bal - prop_B_bal)

if mejora_observada >= umbral_practico:
    print(f"✅ SÍ tiene significancia práctica (>{umbral_practico*100}%)")
else:
    print(f"❌ NO tiene significancia práctica (<{umbral_practico*100}%)")
    
print(f"• Mejora mínima esperada: {umbral_practico*100}%")
print(f"• Mejora observada: {mejora_observada*100:.1f}%")
print(f"• Déficit: {(umbral_practico - mejora_observada)*100:.1f} puntos porcentuales")

=== INTERPRETACIÓN FINAL DE SIGNIFICANCIA PRÁCTICA ===
📊 RESULTADOS CLAVE:
• Diferencia observada: 0.3 puntos porcentuales
• Tamaño del efecto (Cohen's h): 0.0065 (TRIVIAL)
• Significancia estadística: NO (p = 0.775)
• Muestra necesaria: 184,188 usuarios por grupo
• Muestra actual: 3,777 usuarios por grupo

🎯 SIGNIFICANCIA PRÁCTICA:
❌ NO tiene significancia práctica (<5.0%)
• Mejora mínima esperada: 5.0%
• Mejora observada: 0.3%
• Déficit: 4.7 puntos porcentuales


In [46]:
print("\n=== IMPACTO EN EL NEGOCIO ===")

# Calcular impacto monetario hipotético
usuarios_mensuales = 10000  # Ejemplo
ingresos_por_conversion = 50  # Ejemplo: $50 por conversión

# Impacto actual
mejora_conversiones = usuarios_mensuales * mejora_observada
ingresos_adicionales = mejora_conversiones * ingresos_por_conversion

print(f"📈 IMPACTO PROYECTADO (mensual):")
print(f"• Usuarios mensuales estimados: {usuarios_mensuales:,}")
print(f"• Conversiones adicionales: {mejora_conversiones:.0f}")
print(f"• Ingresos adicionales: ${ingresos_adicionales:,.0f}")

# Costo de implementación vs beneficio
costo_implementacion = 50000  # Ejemplo
meses_recuperacion = costo_implementacion / ingresos_adicionales if ingresos_adicionales > 0 else float('inf')

print(f"\n💰 ANÁLISIS COSTO-BENEFICIO:")
print(f"• Costo estimado de implementación: ${costo_implementacion:,}")
if meses_recuperacion != float('inf'):
    print(f"• Tiempo de recuperación: {meses_recuperacion:.1f} meses")
else:
    print("• Tiempo de recuperación: NUNCA (no hay mejora)")


=== IMPACTO EN EL NEGOCIO ===
📈 IMPACTO PROYECTADO (mensual):
• Usuarios mensuales estimados: 10,000
• Conversiones adicionales: 33
• Ingresos adicionales: $1,632

💰 ANÁLISIS COSTO-BENEFICIO:
• Costo estimado de implementación: $50,000
• Tiempo de recuperación: 30.6 meses


### Resultados clave del análisis

Significancia Estadística
* Estadístico z: 0.2862
* Valor p: 0.775 (> 0.05)
* Conclusión: NO hay diferencia estadísticamente significativa

Esto significa que la diferencia del 0.3% entre grupos podría deberse simplemente al azar, **SI hay** usuarios en ambas muestras, aunque el número de eventos por usuario no está distribuido equitativamente entre las muestras en un inicio.

Significancia Práctica

* Diferencia observada: 0.3 puntos porcentuales
* Tamaño del efecto (Cohen's h): 0.0065 (TRIVIAL)
* Mejora esperada: 5% mínimo
* Mejora real: 0.3%

**Problemas Críticos que se identificaron**

* Contaminación de datos: 441 usuarios duplicados en ambos grupos
* Desbalance severo: Grupo A (56.6%) vs Grupo B (43.4%)
* Anomalías en conversión: Más usuarios compraron que agregaron al carrito


# Conclusión

Problemas Críticos Identificados
1. Contaminación de Datos Severa:

* 441 usuarios presentes en ambos grupos (3.2% de contaminación)
* Esto viola completamente la independencia entre grupos
* Hace que cualquier comparación directa sea inválida

2. Desbalance Significativo:

* Grupo A: 8,214 usuarios (56.6%)
* Grupo B: 6,311 usuarios (43.4%)
* Diferencia: 1,903 usuarios (13.2% de desbalance)

3. Distribución Desigual de Actividad:

* Grupo A: Promedio 4.3 eventos, Mediana 3.0
* Grupo B: Promedio 3.2 eventos, Mediana 0.0 (muy preocupante)
* Sugiere que muchos usuarios del Grupo B no tuvieron actividad

4. Anomalías en el Embudo:

* Conversión Cart → Purchase > 100% (más usuarios compraron que agregaron al carrito)
* Indica flujos de compra alternativos (compra directa sin carrito)

**Resultados de la Prueba A/B**

Estrategia de Rescate Implementada

Datos Balanceados:

* Se eliminaron usuarios duplicados
* Se crearon muestras balanceadas de 5,880 usuarios por grupo
* Todas las métricas fueron recalculadas con datos limpios

**Resultados Finales**
Conversión Page → Purchase:

* Grupo A (Control): 50.9%
* Grupo B (Tratamiento): 50.5%
* Diferencia: 0.3 puntos porcentuales

Prueba Z:

* Estadístico z: 0.2862
* Valor p: 0.775 (> 0.05)

Conclusión: NO significativo estadísticamente

**Significancia Práctica**

Tamaño del Efecto:

* Cohen's h: 0.0065 (TRIVIAL)
* Muy por debajo del umbral de significancia práctica

Impacto en el Negocio:

* Mejora esperada: 5% mínimo
* Mejora observada: 0.3%
* Déficit: 4.7 puntos porcentuales

**Limitaciones del Estudio**

Poder Estadístico Insuficiente:

* Muestra actual: 3,777 usuarios por grupo
* Muestra necesaria: 184,188 usuarios por grupo
* La muestra es solo el 2% de lo necesario para detectar esta diferencia
  
**Conclusión Final**

La prueba A/B NO fue exitosa por:

* Problemas de diseño: Contaminación y desbalance invalidan los supuestos básicos
* Sin significancia estadística: p = 0.775 (> 0.05)
* Sin significancia práctica: Mejora trivial del 0.3% vs 5% esperado
* Muestra insuficiente: Se necesita 50x más usuarios para detectar diferencias reales